packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts

```diff
@@ -249,6 +253,157 @@ export class Toolkit extends CloudAssemblySourceBuilder implements AsyncDisposab
     return new IdentityCloudAssemblySource(assembly.assembly);
   }
 
+  /**
+   * Diff Action
+   */
+  public async diff(cx: ICloudAssemblySource, options: DiffOptions): Promise<TemplateDiff> {
+    const ioHelper = asIoHelper(this.ioHost, 'diff');
+    const assembly = await assemblyFromSource(cx);
+    const selectStacks = options.stacks ?? ALL_STACKS;
+    const stackCollection = assembly.selectStacksV2(selectStacks);
+
+    const deployments = await this.deploymentsForAction('diff');
+
+    const strict = !!options.strict;
+    const contextLines = options.contextLines || 3;
+
+    let diffs = 0;
+    let fullDiff;
+    const parameterMap = buildParameterMap(options.parameters?.parameters);
+
+    if (options.templatePath !== undefined) {
+      // Compare single stack against fixed template
+      if (stackCollection.stackCount !== 1) {
+        throw new ToolkitError(
+          'Can only select one stack when comparing to fixed template. Use --exclusively to avoid selecting multiple stacks.',
+        );
+      }
+
+      if (!(await fs.pathExists(options.templatePath))) {
+        throw new ToolkitError(`There is no file at ${options.templatePath}`);
+      }
+
+      const file = fs.readFileSync(options.templatePath).toString();
+      const template = deserializeStructure(file);
+      if (options.securityOnly) {
```

Comment:
All this smells like helper functions to me. It's so hard to read at the moment.

Reply:
ack. i've decided to do a subsequent PR to refactor the code here